src/EnergyPlus/AirflowNetwork/src/Elements.cpp

```diff
@@ -3690,6 +3681,1015 @@ namespace AirflowNetwork {
+            while (true) {
+                if (hghtsFR > zStF(lF + 1)) {
+                    if (lF > 2) break;
+                    ++lF;
+                }
+                if (hghtsFR <= zStF(lF + 1)) break;
+            }
+
+            while (true) {
+                if (hghtsTR > zStT(lT + 1)) {
+                    ++lT;
+                }
+                if (hghtsTR <= zStT(lT + 1)) break;
+            }
+
+            delzF = hghtsF(i) - zStF(lF);
+            delzT = hghtsT(i) - zStT(lT);
+
+            RhoProfF(i + Pprof) = RhoStF(lF) + BetaF(lF) * delzF;
+            RhoProfT(i + Pprof) = RhoStT(lT) + BetaT(lT) * delzT;
+
+            DpProf(i + Pprof) = DpF(lF) - DpT(lT) - G * (RhoStF(lF) * delzF + BetaF(lF) * pow_2(delzF) / 2.0) +
+                                G * (RhoStT(lT) * delzT + BetaT(lT) * pow_2(delzT) / 2.0);
+        }
+    }
+
+    void DetailedOpeningSolver::pstack(EnergyPlusData &state, std::vector<AirflowNetwork::AirState> &props, Array1D<Real64> &pz)
+    {
+
+        // SUBROUTINE INFORMATION:
+        //       AUTHOR         Lixing Gu
+        //       DATE WRITTEN   Oct. 2005
+        //       MODIFIED       na
+        //       RE-ENGINEERED  This subroutine is revised based on PresProfile subroutine from COMIS
+
+        // PURPOSE OF THIS SUBROUTINE:
+        // This subroutine calculates the stack pressures for a link between two zones
+
+        // METHODOLOGY EMPLOYED:
+        // na
+
+        // REFERENCES:
+        // Helmut E. Feustel and Alison Rayner-Hooson, "COMIS Fundamentals," LBL-28560,
+        // Lawrence Berkeley National Laboratory, Berkeley, CA, May 1990
+
+        // USE STATEMENTS:
+        // Locals
+        // SUBROUTINE ARGUMENT DEFINITIONS:
+        // na
+
+        // SUBROUTINE PARAMETER DEFINITIONS:
+        Real64 constexpr PSea(101325.0);
+
+        // INTERFACE BLOCK SPECIFICATIONS
+        // na
+
+        // DERIVED TYPE DEFINITIONS
+        // na
+
+        // SUBROUTINE LOCAL VARIABLE DECLARATIONS:
+        //      REAL(r64) RhoOut ! air density outside [kg/m3]
+        Real64 G;     // gravity field strength [N/kg]
+        Real64 RhoL1; // Air density [kg/m3]
+        Real64 RhoL2;
+        Real64 Pbz;                                                // Pbarom at entrance level [Pa]
+        Array2D<Real64> RhoDrL(state.afn->NumOfLinksMultiZone, 2); // dry air density on both sides of the link [kg/m3]
+        Real64 TempL1;                                             // Temp in From and To zone at link level [C]
+        Real64 TempL2;
+        //      REAL(r64) Tout ! outside temperature [C]
+        Real64 Xhl1; // Humidity in From and To zone at link level [kg/kg]
+        Real64 Xhl2;
+        //      REAL(r64) Xhout ! outside humidity [kg/kg]
+        Array1D<Real64> Hfl(state.afn->NumOfLinksMultiZone); // Own height factor for large (slanted) openings
+        int Nl;                                              // number of links
+
+        Array1D<Real64> DpF(2);
```

Comment:
All of these would ideally be `std::array` so that you are not dynamically allocating them on every call. You could make it `std::array<Real64, 3>` if you don't want to deal with the 0-base thing right now. You can also leave as `Array1D` but make `static` so that it is allocated only once (note, this will make the function thread- and reentry- unsafe).

Reply:
I think this probably could be just the two numbers, it does kind of get looped over but in a very F77 way, there's a refactor needed.